pkg/controller/controller.go

```diff
@@ -1081,47 +1118,108 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+// TODO: This function should eventually be replaced by a common kubernetes library.
```

Comment:
For now this is a pretty important function from a security perspective. Can we move this out into a util file and have extensive unit tests for it?